### ISSUE TITLE
GitHub Actions: Automatic response to feature branch issues

### ISSUE THREAD

**Issue by Aveline-art** [hackforla/website#1940]
### Overview
We need to create a GHA to post a comment on any issue that gets a label `feature: feature branch` with details about what the developer should do.  If the label gets removed, the message should be removed.  This should be a new job that runs in parallel with the other jobs under the "[Issue Trigger](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/issue-trigger.yml)" action.

### Action Items
- [x] Trigger: On event `issues`: activity `opened` if the issue has the `feature: feature branch` OR activity `labeled` with the label `feature: feature branch`
    - [x] Add a comment to the issue:
```
Hi @[REPLACE WITH AUTHOR]
Based on the `feature: feature branch` label, this issue should target a feature branch.  Please consult the instructions on [working off of a feature branch](https://github.com/hackforla/website/wiki/How-to-work-off-of-a-feature-branch)
```
- [x] Trigger: On event `issues`: activity `unlabeled` - if the label `feature: feature branch` is removed
   - [x] Hide (as "outdated") the comment described above

### Checks

- [x] When the `feature: feature branch` label is added during issue creation, the comment is added
- [x] When the `feature: feature branch` label is applied after issue creation, the comment is added
- [x] When a **different label** is added during issue creation, the additional text is **not** added.
- [x] When the `feature: feature branch` label is removed, the comment is hidden
- [x] When a **different label** is removed, the comment is **not** hidden

### Resources/Instructions
Never done GitHub actions? [Start here!](https://docs.github.com/en/actions)

Note that you might want to do something outside the scope of the above psudo-code. If so, be sure to leave comments in your PR or this issue that justifies your reasoning. If you feel you need guidance, be sure to reach out! We cannot foresee whether this issue is solvable, or what hard decisions have to be made, but we would love to hear and help you!

#### Additional resources:
[Events that trigger workflows](https://docs.github.com/en/actions/reference/events-that-trigger-workflows)
[Workflow syntax for GitHub Actions](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions)
[actions/github-script](https://github.com/actions/github-script)
[GitHub RESTAPI](https://docs.github.com/en/rest)
[GitHub GraphQL](https://docs.github.com/en/graphql)
Bonnie test text to remove
Bonnie second test

**Comment by Aveline-art** [comment]
@ExperimentsInHonesty @Sihemgourou 

I am not sure if I should add or edit the labels, but can either of you add the "Feature Branch" label or something akin to that, with a brief description, perhaps, "This issue involves rapid-development of a new feature". Once that is done, this issue can be moved to the backlog with the "Dependency" label removed.

**Comment by ExperimentsInHonesty** [comment]
Says dependency on label but no dependency listed, needs to be reviewed

**Comment by Aveline-art** [comment]
Dependency label is removed. This issue still needs a "Feature Branch" label to be created to move out of the Ice Box.

**Comment by ExperimentsInHonesty** [comment]
new label has been created.

**Comment by ExperimentsInHonesty** [comment]
## This is an old comment and is being left so that if anyone wants to know the history, they can look at it.  But the instructions in the top are the most current instructions.  

@Aveline-art @Zak234 were discussing this issue.  And we have completed this feasibility review.   See additional notes at bottom as well. Please decide if you want this issue to go forward and how.

#### Feasibility Review
- [ ]  When the "Feature Branch" label is added during issue creation, the additional text is added.
- Feasibility: Low for adding things to the top comment on an issue by adding label, not by editing because github does not the top portion of an issue (where original issue is described) as a comment
  - Example: 
this link is the top of this issue https://github.com/hackforla/website/issues/1940#issue-944684210
this link is the comment prior to this one: https://github.com/hackforla/website/issues/1940#issuecomment-898241602
You can see that one has the word comment and the does not.
- Alternative is setting up a different template that automatically has instructions and adds the label.

- [ ]  When a different label is added during issue creation, the additional text is not added.
Feasibility:  high

- [ ]  When the "Feature Branch" label is added after issue creation, the additional text is added.
- Feasibility: Low for adding things to the top comment on an issue by adding label, not by editing because github does not the top portion of an issue (where original issue is described) as a comment
  - Example: 
this link is the top of this issue https://github.com/hackforla/website/issues/1940#issue-944684210
this link is the comment prior to this one: https://github.com/hackforla/website/issues/1940#issuecomment-898241602
You can see that one has the word comment and the does not.
- Alternative is adding it to a comment

- [ ]  When a different label is added after issue creation, the additional text is not added.
   - Feasibility:  high

 - [ ] When the "Feature Branch" label is removed, the additional text is removed.
   - Feasibility: Low if its part of the top portion
   - Feasibility: High if it is its own comment
 
#### Additional options fpr research
When a person open a pr, it could check to see if the issue mentioned has a feature branch label and if it does it could look at the other feature label to see what the feature branch should be called and it could check to see if the commit was made to that branch name.  and if it was not it could give an automated response to the person telling them to remove the pr and submit on the correct branch.  This would require us naming feature branches with the exact name of the feature label.

**Comment by ExperimentsInHonesty** [comment]
@Aveline-art @Zak234 Also forgot to mention that we checked the edit revision history feature of github and you can delete the history but not roll back the edits, so that avenue was a bust.

**Comment by Aveline-art** [comment]
@ExperimentsInHonesty @Zak234 

Thinking about this further, I believe that there are actually two triggers for this action. I have simplified the issue, and hopefully it is a lot more doable. [This action ](https://github.com/hackforla/website/blob/gh-pages/.github/workflows/add-missing-labels-to-issues.yml)by @macho-catt does most of the tasks required for this action (read the labels of a newly created issue, perform an action based on the labels). The only problem to solve would be to edit the top comment, and [this api ](https://docs.github.com/en/rest/reference/issues#update-an-issue)should suffice.

The module @Zak234 was exploring should also suffice as well for this issue after my edits.

**Comment by Zak234** [comment]
Most of the action is finished, the only thing left is using REST API to update the comment with the proper text. I'm having difficulty getting the API to work as I've never used one before. I don't know when it will be finished due to school and time availability.

**Comment by Zak234** [comment]
Don't have time to continue working on this issue due to school. For anyone who picks this up, the most promising solution seemed to be either using REST API  using Update Issue or to add a separate comment with the notes on feature-branches though you would have to convince leadership on the second option.

**Comment by maryjng** [comment]
Availability: 7 - 10 PM EST
ETA: 3/13/2023

**Comment by jdingeman** [comment]
Hi @maryjng - I see you removed your assignment. Could you please provide us with an update that led you to unassigning?

**Comment by jdingeman** [comment]
Spoke with dev. Dev has time constraints and cannot work on issue at this time. Sending this back to the prioritized backlog.

**Comment by arcan9** [comment]
Availability: 2-5pm PST
ETA: EOD 04/18/23

**Comment by arcan9** [comment]
@Aveline-art @ExperimentsInHonesty 

Hi, I'm working on this issue and I just wanted to ask for clarification -- 

1) Aveline-art, you linked this [Github API](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#update-an-issue) for updating an issue. Looks like there are two methods: cURL and JavaScript. I'm assuming we can use cURL? Because the JS method needs the Octokit library dependency, it seems.

2) It looks like a token is needed for auth. Where would I find this token?
```
// cURL
-H "Authorization: Bearer <YOUR-TOKEN>"\
```
EDIT: questions above have been answered in Slack

Thank you!

**Comment by arcan9** [comment]
Another question:

I decided to use the JavaScript method which requires the use of the Octokit library. Here's the code block for a PATCH request according to [the API](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#update-an-issue)
```
// Octokit.js
// https://github.com/octokit/core.js#readme
const octokit = new Octokit({
  auth: 'YOUR-TOKEN'
})

await octokit.request('PATCH /repos/{owner}/{repo}/issues/{issue_number}', {
  owner: 'OWNER',
  repo: 'REPO',
  issue_number: 'ISSUE_NUMBER',
  title: 'Found a bug',
  body: 'I\'m having a problem with this.',
  assignees: [
    'octocat'
  ],
  milestone: 1,
  state: 'open',
  labels: [
    'bug'
  ],
  headers: {
    'X-GitHub-Api-Version': '2022-11-28'
  }
})
```
It says the Octokit library is required. How do I download this library? First I thought I needed to do:
`npm install Octokit`

But when I do that, it generates a new `package.json` in the root folder of the entire project. How do I properly install Octokit? Or should it already be installed and I don't have to worry about it?

**Comment by Jaretzbalba** [comment]
@arcan9 It looks like you're currently working on this issue, so I'm going to move this over to the `in progress` column so that's there no confusion! Please feel free to change it back if otherwise.

**Comment by arcan9** [comment]
Update on my work on this issue: I'm deciding to unassign myself due to some personal life stuff. Therefore I'm taking a break. But I'll leave a couple of notes that might help the next developer who'll be working on this issue.

  - Per the instructions of this issue, I added a new job under the "Issue Trigger" GitHub action found in `/.github/workflows/issue-trigger.yml`
  - This job is intended to append new additional text when an issue is created AND labeled "Feature Branch"
  -  Then I created a new script that contains the appended text. This script will only run if certain conditions are met.
  - I'm using the API described in this comment: https://github.com/hackforla/website/issues/1940#issuecomment-901350940
  - I created all that's required including the Github Token/Secret (I followed the instructions [here](https://github.com/hackforla/website/wiki/Hack-for-LA's-GitHub-Actions#tip-7-using-personal-access-tokens-to-test-in-your-own-project-board)).
  
**Testing in my own repo:** 

- I created a secret token
- I added that secret to the `yml` and `js` files
- I opened an issue and labeled it `Feature Branch` and `Feature: Feature Branch` to see which one of these would trigger the job
- Unfortunately, nothing happens

Hopefully the next dev who works on this will be able to solve the problem. To see what my code looked like, [here's the PR I created](https://github.com/hackforla/website/pull/4496) (now closed).

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty  This issue as written is not complete and I'm not convinced that it is feasible or even necessary.   In your comment https://github.com/hackforla/website/issues/1940#issuecomment-901268976 you raised several significant objections, including the fact that if the `feature branch` label were removed, it might not be feasible to remove the feature branch instructions if they are appended to the top portion of the issue.  You mentioned that feasibility would be higher if the instructions were placed in their own comment but it doesn't look as if that idea was pursued, perhaps because it would not be as effective.  In fact the issue as written does not address the situation in which the `feature branch` label is removed.

**Comment by ExperimentsInHonesty** [comment]
I am adding this to the next agenda, so we can revisit.  I also think we could have an issue template for feature branch issues, that has this info built into it if we need it.

**Comment by ExperimentsInHonesty** [comment]
Please rewrite the action items on this issue to create a GHA that will post a comment if the feature: feature branch label is used and remove the comment if the label gets removed.

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty Those changes have been incorporated into the issue.

**Comment by naveenmallemala5** [comment]
Availability: Mon - Friday 9:00 AM - 6:00 PM
ETA: 02/08/2024

**Comment by naveenmallemala5** [comment]
Availability: Mon - Friday 9:00 AM - 6:00 PM
ETA: 02/23/2024

**Comment by LRenDO** [comment]
Hi @naveenmallemala5! Thanks for working on this issue. If you're still working, please add an update. If you're not able to continue to working on it, leave us a comment to let us know.  If we don't hear from you soon, we will remove you from the issue. Thanks!

**Comment by LRenDO** [comment]
Hi @naveenmallemala5! Since we haven't received an update from you, I'm taking you off the issue and moving it for re-prioritization. Feel free to take on another issue, if you become available.

**Comment by partapparam** [comment]
i. Availability: 6-9pm Wed and Thursday. 12-6 pm Sunday.
ii. ETA: 4/26

**Comment by ExperimentsInHonesty** [comment]
@partapparam I will move this issue the in progress column, but in future, you should do that when you self assign.

**Comment by partapparam** [comment]
whoops, sorry. thanks @ExperimentsInHonesty

**Comment by partapparam** [comment]
**Update**
i. Availability: 3-6pm Friday, 9-3pm Sat.
ii. ETA: 4/27
